How many late days are you using for this assignment? 

In [40]:
# Imports and pip installations (if needed)
%pip install pandas
%pip install pandas scikit-learn
import pandas as pd
from sklearn.preprocessing import MinMaxScaler



Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 23.1.2 -> 25.0.1
[notice] To update, run: pip3.11 install --upgrade pip
Note: you may need to restart the kernel to use updated packages.
Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.1/11.1 MB 31.3 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 22.4/22.4 MB 26.1 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 301.8/301.8 kB 20.4 MB/s eta 0:00:00

[notice] A new release of pip is available: 23.1.2 -> 25.0.1
[notice] To update, run: pip3.11 install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


# Part 1: Load the dataset

In [8]:
# Load the given datasets
chronic_numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")
chronic_categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")

# Print the data
print("Numerical Dataset:")
print(chronic_numerical.head())

print("\nCategorical Dataset:")
print(chronic_categorical.head())

Numerical Dataset:
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  8200.0   
4     240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN     NaN   

   rbcc  
0   6.4  
1   5.5  
2   5.7  
3   4.5  
4   NaN  

Categorical Dataset:
   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0  

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The dataset provides information on the patients that help in predicting whether a patient has chronic kidney disease. Rows represent one patient and the columns are attributes features that describe their medical measurements and test results. The label is whether or not the patient has CKD. The dataset includes both numerical: age, bu, sc and categorical features: rbc, htn, dm.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [18]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicate_numerical = chronic_numerical[chronic_numerical.duplicated()]
total_duplicate_numerical = duplicate_numerical.shape[0]
print(f"Total number of duplicate rows in numerical dataset: {total_duplicate_numerical}")

# Drop these duplicate rows
chronic_numerical = chronic_numerical.drop_duplicates()
print("Duplicate rows in numerical dataset dropped.")

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicate_categorical = chronic_categorical[chronic_categorical.duplicated()]
total_duplicate_categorical = duplicate_categorical.shape[0]
print(f"Total number of duplicate rows in categorical dataset: {total_duplicate_categorical}")

chronic_categorical = chronic_categorical.drop_duplicates()
print("Duplicate rows in categorical dataset dropped.")

Total number of duplicate rows in numerical dataset: 0
Duplicate rows in numerical dataset dropped.
Total number of duplicate rows in categorical dataset: 0
Duplicate rows in categorical dataset dropped.


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No because some unique ID's might be present in one dataset and not in the other.

In [21]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_data = pd.merge(chronic_numerical, chronic_categorical, on="unique_id", how="inner")

#Print the combined dataset
print("Combined Dataset:")
print(merged_data.head())

Combined Dataset:
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3     764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
4     240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   

       pc         pcc          ba  htn   dm cad appet   pe ane  Target  
0  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
1  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
2  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
3  normal  notpresent  notpresent  yes  yes  no  poor  yes  no     ckd  
4  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  

[5 rows x 25 columns]


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [32]:
# Calculate the percentage of rows that contain atleast one missing value
missing_rows = merged_data[merged_data.isnull().any(axis=1)]
total_rows = merged_data.shape[0]
total_missing_rows = missing_rows.shape[0]
percentage_missing_rows = (total_missing_rows / total_rows) * 100
# Print %
print(f"Percentage of rows that contain atleast one missing value: {percentage_missing_rows:.2f}%")
# Drop these rows from the dataset
merged_data_clean = merged_data.dropna()
# Print the Dataset
print("Dataset after dropping rows with missing values:")
print(merged_data_clean.head())

Percentage of rows that contain atleast one missing value: 60.77%
Dataset after dropping rows with missing values:
    unique_id   age    bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0      203694  38.0  80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1      938027  43.0  60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2      421471  37.0  60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
7      109053  64.0  60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
13     349892  59.0  70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   

          pc         pcc          ba  htn   dm  cad appet  pe ane  Target  
0     normal  notpresent  notpresent   no   no   no  good  no  no  notckd  
1     normal  notpresent  notpresent   no   no   no  good  no  no  notckd  
2     normal  notpresent  notpresent   no   no   no  good  no  no  notckd  
7     normal  notpresent  notpresent   no   no   no  good  no  no  notckd  
13  abnormal  notpresent  notpresent  yes  yes  

## Part 3.4: Sort the dataset according to the Labels

In [34]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
merged_data_sorted = merged_data_clean.sort_values(by='Target').reset_index(drop=True)
# Print the dataset
print("Dataset sorted according to 'Target' column:")
print(merged_data_sorted.head())

Dataset sorted according to 'Target' column:
   unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1     137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2     484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3     343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4     532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   

         pc         pcc          ba  htn   dm  cad appet   pe  ane Target  
0  abnormal     present     present   no   no   no  good   no  yes    ckd  
1  abnormal     present  notpresent  yes  yes  yes  poor   no   no    ckd  
2  abnormal     present  notpresent  yes  yes   no  poor   no  yes    ckd  
3  abnormal  notpresent  notpresent  yes  yes  yes  poor  yes  yes    ckd  
4    normal  notpresent  notpresent   no   no   no  good   no   no    ckd  

[5 rows x 25 columns]


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [35]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_data = pd.get_dummies(merged_data, columns=categorical_columns, drop_first=False)

# Print the dataset
print(encoded_data)


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
385     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0  ...   
386     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
389     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0  ...   

     cad_yes  appet_good  appet_poor  pe_no  pe_yes

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: If we replace multiple binary columns with a single column representing a binary variable, the downstream modeling may not always be equivalent. If the categorical feature can take more than two values using a single column instead of one-hot encoding can lead to loss of information or improper interpretation of the variable.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [36]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for col in numerical_columns:
    mean = encoded_data[col].mean()
    std = encoded_data[col].std()
    
    encoded_data = encoded_data[(encoded_data[col] >= mean - 3 * std) & (encoded_data[col] <= mean + 3 * std)]

# Print the dataset
print(encoded_data)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
5       992643  35.0   70.0   82.0   36.0  1.1  150.0  3.5  14.5  52.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
381     462892  36.0   80.0   85.0   16.0  1.1  142.0  4.1  15.6  44.0  ...   
383     838360  71.0   90.0   88.0   80.0  4.4  139.0  5.7  11.3  33.0  ...   
386     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   

     cad_yes  appet_good  appet_poor  pe_no  pe_yes

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [41]:
# Normalize the all Numerical Attributes in the dataset.
scaler = MinMaxScaler()
encoded_data[numerical_columns] = scaler.fit_transform(encoded_data[numerical_columns])
# Print the dataset
print("Normalized Dataset:")
print(encoded_data.head())

Normalized Dataset:
   unique_id       age    bp       bgr        bu        sc       sod  \
0     203694  0.380952  0.50  0.093548  0.058824  0.008696  0.900000   
1     938027  0.440476  0.00  0.122581  0.098039  0.052174  0.800000   
2     421471  0.369048  0.00  0.132258  0.163399  0.034783  0.500000   
3     764115  0.761905  0.75  0.238710  0.751634  0.313043  0.533333   
5     992643  0.345238  0.25  0.038710  0.169935  0.060870  1.000000   

        pot      hemo       pcv  ...  cad_yes  appet_good  appet_poor  pe_no  \
0  0.256410  0.655738  0.736842  ...    False        True       False   True   
1  0.641026  1.000000  0.710526  ...    False        True       False   True   
2  0.410256  0.868852  0.894737  ...    False        True       False   True   
3  0.538462  0.524590  0.552632  ...    False       False        True  False   
5  0.256410  0.729508  0.947368  ...    False        True       False   True   

   pe_yes  ane_no  ane_yes  Target_ckd  Target_ckd\t  Target_notck

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: 

In [53]:
#Remove that column
final_data = encoded_data.drop(columns=['Target_notckd'], errors='ignore')

final_data['htn'] = final_data[['htn_no', 'htn_yes']].idxmax(axis=1).str.replace('htn_', '')
final_data['dm'] = final_data[['dm_no', 'dm_yes']].idxmax(axis=1).str.replace('dm_', '')
final_data['cad'] = final_data[['cad_no', 'cad_yes']].idxmax(axis=1).str.replace('cad_', '')
final_data['appet'] = final_data[['appet_good', 'appet_poor']].idxmax(axis=1).str.replace('appet_', '')
final_data['pe'] = final_data[['pe_no', 'pe_yes']].idxmax(axis=1).str.replace('pe_', '')
final_data['ane'] = final_data[['ane_no', 'ane_yes']].idxmax(axis=1).str.replace('ane_', '')

# Drop old one-hot encoded columns
one_hot_columns = ['htn_no', 'htn_yes', 'dm_no', 'dm_yes', 'cad_no', 'cad_yes', 
                   'appet_good', 'appet_poor', 'pe_no', 'pe_yes', 'ane_no', 'ane_yes']
final_data = final_data.drop(columns=one_hot_columns, errors='ignore')

# Display cleaned DataFrame
print(final_data.head())

# Print the dataset
print(final_data.head())

   unique_id       age    bp       bgr        bu        sc       sod  \
0     203694  0.380952  0.50  0.093548  0.058824  0.008696  0.900000   
1     938027  0.440476  0.00  0.122581  0.098039  0.052174  0.800000   
2     421471  0.369048  0.00  0.132258  0.163399  0.034783  0.500000   
3     764115  0.761905  0.75  0.238710  0.751634  0.313043  0.533333   
5     992643  0.345238  0.25  0.038710  0.169935  0.060870  1.000000   

        pot      hemo       pcv  ...  ba_present  dm_\tno  Target_ckd  \
0  0.256410  0.655738  0.736842  ...       False    False       False   
1  0.641026  1.000000  0.710526  ...       False    False       False   
2  0.410256  0.868852  0.894737  ...       False    False       False   
3  0.538462  0.524590  0.552632  ...       False    False        True   
5  0.256410  0.729508  0.947368  ...       False    False       False   

   Target_ckd\t  htn   dm  cad  appet   pe  ane  
0         False   no   no   no   good   no   no  
1         False   no   no   

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [54]:
csv_file_path = "chronic_kidney_disease_cleaned_data.csv"
json_file_path = "chronic_kidney_disease_cleaned_data.json"

# Export the DataFrame to a new CSV file
final_data.to_csv(csv_file_path, index=False)

# Export the DataFrame to a new JSON file
final_data.to_json(json_file_path, orient='records', lines=True)

# Print confirmation messages
print(f"DataFrame exported to {csv_file_path}")
print(f"DataFrame exported to {json_file_path}")

DataFrame exported to chronic_kidney_disease_cleaned_data.csv
DataFrame exported to chronic_kidney_disease_cleaned_data.json


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [55]:
top_15_rows = final_data.head(15)
print(top_15_rows)

    unique_id       age    bp       bgr        bu        sc       sod  \
0      203694  0.380952  0.50  0.093548  0.058824  0.008696  0.900000   
1      938027  0.440476  0.00  0.122581  0.098039  0.052174  0.800000   
2      421471  0.369048  0.00  0.132258  0.163399  0.034783  0.500000   
3      764115  0.761905  0.75  0.238710  0.751634  0.313043  0.533333   
5      992643  0.345238  0.25  0.038710  0.169935  0.060870  1.000000   
7      109053  0.690476  0.00  0.116129  0.111111  0.026087  1.000000   
8      352995  0.321429  0.00  0.193548  0.202614  0.043478  0.700000   
10     261025  0.666667  0.75  0.319355  0.248366  0.173913  0.600000   
17     630177  0.619048  0.50  0.196774  0.052288  0.060870  0.700000   
18     741276  0.202381  0.00  0.080645  0.091503  0.034783  0.833333   
19     546010  0.511905  0.50  0.167742  0.143791  0.069565  0.633333   
20     460465  0.488095  0.00  0.125806  0.098039  0.060870  0.700000   
22     965166  0.500000  0.50  0.016129  0.078431  

| unique_id | age  | bp  | bgr  | bu  | sc  | sod  | pot  | hemo | pcv  | wbcc   | rbcc |
|-----------|------|-----|------|-----|-----|------|------|------|------|--------|------|
| 203694    | 38.0 | 80.0 | 99.0  | 19.0 | 0.5 | 147.0 | 3.5  | 13.6 | 44.0 | 7300.0 | 6.4  |
| 938027    | 43.0 | 60.0 | 108.0 | 25.0 | 1.0 | 144.0 | 5.0  | 17.8 | 43.0 | 7200.0 | 5.5  |
| 421471    | 37.0 | 60.0 | 111.0 | 35.0 | 0.8 | 135.0 | 4.1  | 16.2 | 50.0 | 5500.0 | 5.7  |
| 764115    | 70.0 | 90.0 | 144.0 | 125.0 | 4.0 | 136.0 | 4.6  | 12.0 | 37.0 | 8200.0 | 4.5  |
| 240975    | 47.0 | 80.0 | 95.0  | 35.0 | 0.9 | 140.0 | 4.1  |      |      |        |      |
| 992643    | 35.0 | 70.0 | 82.0  | 36.0 | 1.1 | 150.0 | 3.5  | 14.5 | 52.0 | 9400.0 | 6.1  |
| 573400    | 60.0 | 80.0 |       |      |     |      |      |      |      |        |      |
| 109053    | 64.0 | 60.0 | 106.0 | 27.0 | 0.7 | 150.0 | 3.3  | 14.4 | 42.0 | 8100.0 | 4.7  |
| 352995    | 33.0 | 60.0 | 130.0 | 41.0 | 0.9 | 141.0 | 4.4  | 15.5 | 52.0 | 4300.0 | 5.8  |
| 395450    | 65.0 | 70.0 | 203.0 | 46.0 | 1.4 |      |      | 11.4 | 36.0 | 5000.0 | 4.1  |
| 261025    | 62.0 | 90.0 | 169.0 | 48.0 | 2.4 | 138.0 | 2.9  | 13.4 | 47.0 | 11000.0 | 6.1  |
| 389919    | 61.0 | 90.0 |       |      |     |      |      |      |      | 9800.0 |      |
| 210452    | 72.0 | 90.0 | 124.0 | 53.0 | 2.3 |      |      | 11.9 | 39.0 |        |      |
| 349892    | 59.0 | 70.0 | 424.0 | 55.0 | 1.7 | 138.0 | 4.5  | 12.6 | 37.0 | 10200.0 | 4.1  |


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [56]:
# Code to rename all the columns in the dataset
# Create a dictionary mapping abbreviated column names to descriptive names
rename_dict = {
    'age': 'Age',
    'bp': 'Blood_Pressure',
    'bgr': 'Blood_Glucose_Random',
    'bu': 'Blood_Urea',
    'sc': 'Serum_Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed_Cell_Volume',
    'wbcc': 'WBC_Count',
    'rbcc': 'RBC_Count',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc_abnormal': 'RBC_Status_Abnormal',
    'rbc_normal': 'RBC_Status_Normal',
    'pc_abnormal': 'Pus_Cell_Abnormal',
    'pc_normal': 'Pus_Cell_Normal',
    'pcc_notpresent': 'Pus_Cell_Clumps_Not_Present',
    'pcc_present': 'Pus_Cell_Clumps_Present',
    'ba_notpresent': 'Bacteria_Not_Present',
    'ba_present': 'Bacteria_Present',
    'htn_no': 'Hypertension_No',
    'htn_yes': 'Hypertension_Yes',
    'dm_no': 'Diabetes_Mellitus_No',
    'dm_yes': 'Diabetes_Mellitus_Yes',
    'cad_no': 'Coronary_Artery_Disease_No',
    'cad_yes': 'Coronary_Artery_Disease_Yes',
    'appet_good': 'Appetite_Good',
    'appet_poor': 'Appetite_Poor',
    'pe_no': 'Pedal_Edema_No',
    'pe_yes': 'Pedal_Edema_Yes',
    'ane_no': 'Anemia_No',
    'ane_yes': 'Anemia_Yes',
    'Target_ckd': 'Target_CKD',
    'Target_notckd': 'Target_Not_CKD'
}

# Apply the renaming to the final dataset (assumed to be in the variable final_data)
final_data_renamed = final_data.rename(columns=rename_dict)

# Print the first few rows of the renamed dataset
print(final_data_renamed.head())


   unique_id       Age  Blood_Pressure  Blood_Glucose_Random  Blood_Urea  \
0     203694  0.380952            0.50              0.093548    0.058824   
1     938027  0.440476            0.00              0.122581    0.098039   
2     421471  0.369048            0.00              0.132258    0.163399   
3     764115  0.761905            0.75              0.238710    0.751634   
5     992643  0.345238            0.25              0.038710    0.169935   

   Serum_Creatinine    Sodium  Potassium  Hemoglobin  Packed_Cell_Volume  ...  \
0          0.008696  0.900000   0.256410    0.655738            0.736842  ...   
1          0.052174  0.800000   0.641026    1.000000            0.710526  ...   
2          0.034783  0.500000   0.410256    0.868852            0.894737  ...   
3          0.313043  0.533333   0.538462    0.524590            0.552632  ...   
5          0.060870  1.000000   0.256410    0.729508            0.947368  ...   

   Bacteria_Present  dm_\tno  Target_CKD  Target_ckd\t  

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [57]:
# Converted SQL to Pandas code
result = final_data.groupby('Target_ckd').size().reset_index(name='count')
print(result)



   Target_ckd  count
0       False    124
1        True     64


*Describe what the above code does here*: The code groups the dataset by the target variable (in this case, using the column name "Target_ckd") and then counts the number of rows (i.e., the number of observations) in each group. It resets the index so that the grouping variable becomes a regular column in the resulting DataFrame, and finally prints the resulting table, which shows each unique target value along with its corresponding count.